### PR TITLE
T13893 registry collection

### PR DESCRIPTION
**Pull request by niden**
Hello!

* Type: new feature | code quality
* Link to issue: #13893 

**In raising this pull request, I confirm the following:**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist
- [x] I wrote some tests for this PR
- [x] I updated the CHANGELOG

Refactored `Registry` to use `Collection`;
Refactored `Session\Bag` to use `Collection`

Thanks

### CODE REVIEW

**Comment by niden** [comment]
cc @sergeyklay

File: phalcon/collection.zep [hunk 1]
```diff
@@ -86,12 +91,16 @@ class Collection implements \ArrayAccess, \Countable, \IteratorAggregate, \JsonS
...
 	 * Get the element from the collection
 	 */
-	public function get(string! element, var defaultValue = null) -> var | bool
+	public function get(string! element, var defaultValue = null, bool insensitive = true) -> var | bool
 	{
 		var value;
 
-		if likely fetch value, this->data[element] {
-			return value;
+		if insensitive {
```
Comment (sergeyklay) [inline phalcon/collection.zep]:
```suggestion
		if likely insensitive {
```

Comment (sergeyklay) [reply phalcon/collection.zep]: You have to use the branch prediction here

File: phalcon/collection.zep [hunk 2]
```diff
@@ -108,44 +117,25 @@ class Collection implements \ArrayAccess, \Countable, \IteratorAggregate, \JsonS
...
-	/**
-	 * Set an element in the collection
-	 */
-	public function set(string! element, var value) -> void
-	{
-		array data;
-
-		let data          = this->data,
-			data[element] = value;
+		if insensitive {
```
Comment (sergeyklay) [inline phalcon/collection.zep]: the same

File: phalcon/collection.zep [hunk 3]
```diff
@@ -108,44 +117,25 @@ class Collection implements \ArrayAccess, \Countable, \IteratorAggregate, \JsonS
...
 
-		this->init(data);
+		return isset this->headers[element];
 	}
 
 	/**
 	 * Initialize internal array
 	 */
-	public function init(array! data = [])
+	public function init(array! data = []) -> void
```
Comment (sergeyklay) [inline phalcon/collection.zep]: How about this?
```suggestion
	internal function init(array! data = []) -> void
```

Comment (sergeyklay) [reply phalcon/collection.zep]: why do we need this in the API part?

File: phalcon/collection.zep [hunk 4]
```diff
@@ -24,13 +24,18 @@ class Collection implements \ArrayAccess, \Countable, \IteratorAggregate, \JsonS
...
 	 * @var array
 	 */
-	private data = [];
+	protected data = [];
 
-	public function __construct(array! data = null) -> void
+	/**
+	 * @var array
+	 */
+	protected headers = [];
```
Comment (sergeyklay) [inline phalcon/collection.zep]: Is it a common class for any other collections, right? Why do we need domain specific headers in the base class? I thought that `data` is enough for data storage. Why do we need `data` and `headers` here?

Comment (sergeyklay) [reply phalcon/collection.zep]: How about HeadersBag or something like this?

**Review by sergeyklay — CHANGES_REQUESTED** [review]

File: phalcon/collection.zep (hunk 3 above)
Comment (niden) [reply phalcon/collection.zep]: I want to offer the option to populate the collection after it was created. For instance you can reuse the same object to populate data from another array that came in later in the code flow..

File: phalcon/collection.zep (hunk 4 above)
Comment (niden) [reply phalcon/collection.zep]: This is not the `headers` as you think of it. This array holds the case insensitive header -> case sensitive header map for faster searching and locating. The name of the internal array/property can be anything./ I just named it `headers`.

I will change it to arrayKeys or something.

**Review by sergeyklay — DISMISSED** [review]
Looks good to me!

**Pull request merged**